### PR TITLE
RBAC Phase 2: remove legacy permission REST endpoints + client methods

### PR DESCRIPTION
**Pull request by PattaraS**
## 🥞 Stacked PR

- [stack/rbac/phase2-simplified](https://github.com/mlflow/mlflow/pull/22855) — ✅ MERGED
  - **[stack/rbac/phase2-legacy-endpoints](https://github.com/mlflow/mlflow/pull/22859)** [[Files changed](https://github.com/mlflow/mlflow/pull/22859/files)] ← you are here
    - [stack/rbac/phase2-store-cleanup](https://github.com/mlflow/mlflow/pull/22861) [[Files changed (incremental)](https://github.com/mlflow/mlflow/pull/22861/files/4fcf48350..6a3555079)]
      - [stack/rbac/phase2-drop-legacy-tables](https://github.com/mlflow/mlflow/pull/23089) [[Files changed (incremental)](https://github.com/mlflow/mlflow/pull/23089/files/6a3555079..31f9973c8)] — DRAFT, graduation, target MLflow 3.X+2

**Stacks on the merged #22855**, now rebased directly onto `master` and merged with the latest `master` to pick up trace-parameterized routing.

---

### What this PR does (revised after review)

The original revision of this PR removed both the legacy resource-permission REST endpoints and the workspace-permission REST endpoints. Per @TomeHirata's review (https://github.com/mlflow/mlflow/pull/22859#discussion_r3198692246), the **24 resource-permission CRUD endpoints stay** so pre-RBAC clients keep working — they're now thin deprecation-warned wrappers over the synthetic-per-user-role grants that #22855's migration writes.

| What | Status | Backing store |
|---|---|---|
| `experiment / registered_model / scorer + 3× gateway` permission CRUD (24 endpoints + 24 `AuthServiceClient` methods) | **KEPT, deprecated** | Synthetic `__user_<id>__` role grants in `role_permissions` |
| `set/delete/list_workspace_permission` + `list_user_workspace_permissions` (4 endpoints + 4 client methods) | **REMOVED** | n/a — semantics changed too much under the simplified model |

Everything else this PR did stays:

- **Filter simplification** — `filter_experiment_ids`, `filter_search_experiments`, `filter_search_logged_models`, `filter_search_registered_models`, `filter_search_model_versions`, and the GraphQL model-version filter all consult `role_permissions` through a new `_role_based_read_predicate(username, resource_type)` helper that folds specific grants, wildcard grants, explicit `NO_PERMISSIONS` denies, and the workspace-vs-default fallback into a single callable.
- **Workspace-permission validators + helpers gone** — `validate_can_list_workspace_permissions`, `validate_can_modify_workspace_permission`, `_workspace_permission` and the six `_workspace_permission_for_*` helpers, the `_has_*_read_access` family, and the parameterized-path matcher that existed solely to route `/workspaces/<workspace_name>/permissions`.

### Deprecation signaling

- Each of the 24 `AuthServiceClient` methods emits a one-shot `FutureWarning` directing callers to `add_role_permission` + `assign_role`. (`FutureWarning` rather than `DeprecationWarning` so it's actually visible to end users — Python suppresses `DeprecationWarning` by default.)
- Each handler also logs a one-time deprecation note per process so server operators see the move signal.

### Review pointers

| File | What changed | Reviewer focus |
|---|---|---|
| `mlflow/server/auth/__init__.py` | New `_role_based_read_predicate` helper; legacy resource-permission handlers preserved as deprecation-logged thin wrappers; workspace-permission handlers + helpers removed. | The predicate is the load-bearing piece. Verify it correctly handles specific grants, wildcard grants, explicit `NO_PERMISSIONS` denies, and the workspace-vs-default fallback. Also verify the merge resolution kept the `TRACE_PARAMETERIZED_BEFORE_REQUEST_VALIDATORS` block from `master`. |
| `mlflow/server/auth/routes.py` | Workspace-permission route constants removed; resource-permission route constants kept under a "Legacy per-resource permission routes" header. | Mechanical. |
| `mlflow/server/auth/client.py` | 4 workspace-permission `AuthServiceClient` methods removed; 24 resource-permission methods kept with `FutureWarning` emission. | Mechanical. |
| `tests/server/auth/auth_test_utils.py` | New `grant_role_permission(tracking_uri, username, resource_type, resource_pattern, permission, workspace=...)` helper — creates a throwaway role + permission + assignment via 3 admin HTTP calls. | Replaces ~30 legacy `POST /permissions/create` setups. |
| `tests/server/auth/test_auth.py`, `test_client.py`, `test_client_workspace.py` | Call sites rewired to `grant_role_permission`. | Mechanical. |
| `tests/server/auth/test_client.py` | `test_legacy_permission_endpoints_remain_registered` (parametrized over the 24 paths × method combos) + `test_legacy_client_methods_emit_deprecation_warning`. | Pins both that the kept endpoints still serve and that the client methods emit the move signal. |
| `docs/api_reference/api_inventory.txt` | Drops the 4 removed workspace-permission method entries. | Mechanical. |

### Behavioral note

In the legacy world, the resource creator had MANAGE on the resource and could delegate permissions to other users via `POST /permissions/create`. In the role world, grant delegation requires workspace-admin or global admin — so the new `grant_role_permission` test helper uses admin credentials. Production deployments that relied on owner-delegation for the now-deprecated endpoints should either elevate their workspace admins or accept that rebinding per-resource grants goes through the role API.

### What remains for a follow-up

The store methods `create_*_permission`, `delete_*_for_*`, and `rename_registered_model_permissions` are still present — they back both the deprecated REST handlers and the internal cascade handlers (`set_can_manage_*_permission`, `delete_*_permissions_cascade`). #22861 was originally going to fold those into a single internal `grant_user_permission(...)` helper and delete them; with the legacy endpoints staying, that PR will need to either keep the per-resource methods (route the deprecated handlers through the new helpers and the per-resource methods become thin shims) or be reshaped accordingly.

### How is this PR tested?

- [x] Full `tests/server/auth/` suite (~464 tests) passes, including the new `test_legacy_permission_endpoints_remain_registered` (24 cases) and `test_legacy_client_methods_emit_deprecation_warning`.
- [x] Lint clean: `uv run ruff check`, `uv run ruff format --check`, `uv run clint`.

### Does this PR require documentation update?

- [x] No.

### Does this PR require updating the [MLflow Skills](https://github.com/mlflow/skills) repository?

- [x] No.

#### Is this a user-facing change?

- [x] Yes — narrowed scope. The 4 workspace-permission REST endpoints and their `AuthServiceClient` methods (`set_workspace_permission`, `delete_workspace_permission`, `list_workspace_permissions`, `list_user_workspace_permissions`) are removed. The 24 per-resource permission CRUD endpoints + matching `AuthServiceClient` methods (experiment / registered_model / scorer / gateway secret/endpoint/model_definition) are **retained** but deprecated; each call now emits a `FutureWarning` directing callers to the role API (`add_role_permission` + `assign_role`).

#### What component(s), interfaces, languages, and integrations does this PR affect?

- [x] `area/server-infra`: MLflow Tracking server, JavaScript dev server
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

#### How should the PR be classified in the release notes?

- [x] `rn/breaking-change` — workspace-permission endpoints removed.

#### Is this PR a critical bugfix or security fix that should go into the next patch release?

- [ ] This PR is critical and needs to be in the next patch release
- [x] This PR can wait for the next minor release

### CODE REVIEW

**Comment by PattaraS** [comment]
/review

---

✅ [Review](https://github.com/mlflow/mlflow/actions/runs/25442867892?pr=22859) completed (344.5s, 42 turns, 17072 tokens, $2.77)
<details><summary>Review Output</summary>

**Review complete for PR #22859 — RBAC Phase 2: remove legacy permission REST endpoints + client methods.**

**Summary:** This is a large, well-structured removal of the pre-RBAC permission API surface. The core new logic — `_role_based_read_predicate` — is sound: it correctly prioritizes specific denies > specific grants > wildcard grants > fallback defaults, and the filter functions are cleanly simplified to delegate to it. The `_user_can_create_in_workspace` and `_validate_can_use_model_definitions_for_create` changes properly preserve the `grant_default_workspace_access` fallback. Test rewiring to `grant_role_permission` is mechanical and consistent. New workspace-aware tests cover the autogrant edge cases.

**One MODERATE finding:** `test_legacy_permission_endpoints_are_removed` checks `/api/2.0/` for 8 paths (scorers, gateway secrets/endpoints/model-definitions) that were actually served at `/api/3.0/`. The test passes but won't catch regression for those v3 endpoints.

PR approved.

</details>

File: mlflow/server/auth/routes.py [hunk 1]
```diff
@@ -21,22 +21,6 @@
 AJAX_DELETE_USER = _get_ajax_path("/mlflow/users/delete")
 LIST_USERS = _get_rest_path("/mlflow/users/list")
 AJAX_LIST_USERS = _get_ajax_path("/mlflow/users/list")
-CREATE_EXPERIMENT_PERMISSION = _get_rest_path("/mlflow/experiments/permissions/create")
```
Comment (TomeHirata) [inline mlflow/server/auth/routes.py:24]: Shouldn't we keep these routes for some time for backward compatibility? I think we can update the permission for the user's synthetic role instead of direct permission records.

Comment (PattaraS) [reply mlflow/server/auth/routes.py:24]: Good call. Pushed `9cc9e323f` which keeps the 24 resource-permission CRUD routes (experiment / registered_model / scorer + the three gateway families) and their `AuthServiceClient` methods. The handlers route through the synthetic per-user role grants written by migration `e5f6a7b8c9d0` — no new storage path, just thin wrappers.

To signal the move:
- Each client method emits a one-shot `FutureWarning` pointing at the role API (`add_role_permission` + `assign_role`).
- Each handler logs a one-time deprecation note per process so server operators see it too.

I left the workspace-permissions endpoints (`set_workspace_permission` / `list_workspace_permissions` / etc.) removed — their semantics changed too much under the simplified model to be worth keeping as compat shims, but happy to revisit if you want those preserved as well.

🤖 *Claude response*

File: mlflow/server/auth/__init__.py [hunk 2]
```diff
@@ -1409,15 +1195,78 @@ def validate_can_view_user_roles():
...
+                # below by the ``readable`` membership check). The simplified RBAC
+                # model rejects wildcard ``NO_PERMISSIONS`` at write time and the
+                # ``e5f6a7b8c9d0`` migration drops legacy wildcard
+                # ``NO_PERMISSIONS`` rows, so this branch is defensive: it pins
+                # the semantics for any row that lands via manual DB edit or a
+                # future relaxation of the validators.
+                wildcard_deny = True
+        elif can_read:
+            readable.add(resource_pattern)
+        else:
```
Comment (TomeHirata) [inline mlflow/server/auth/__init__.py:1224]: nit: I'm not sure if we need this branch. I think it's unrealistic to add "NO_PERMISSIONS" to all users who shouldn't have access. If that's needed, users should make NO_PERMISSIONS the default permission. Also this is inconsistent to other permission resolution that uses max(*permissions).

**Review by TomeHirata — APPROVED** [review]
LGTM with one minor comment!

File: mlflow/server/auth/__init__.py (hunk 2 above)
Comment (PattaraS) [reply mlflow/server/auth/__init__.py:1224]: Good catch — agreed on both points. Pushed `822db537c` which drops both the per-resource and wildcard `NO_PERMISSIONS` deny branches; the predicate is now max-style consistent with the rest of the resolvers:

```python
for resource_pattern, permission in store.list_role_grants_for_user_in_workspace(...):
    if not get_permission(permission).can_read:
        continue
    if resource_pattern == "*":
        wildcard_can_read = True
    else:
        readable.add(resource_pattern)
```

The simplified validators already reject `NO_PERMISSIONS` as a grantable role permission, so the loop's `continue` is defense-only; rebinding the test from `wildcard_no_permissions_blocks_default` to `ignores_no_permissions_grants` to pin the new semantics. Operators that want deny-by-default set `default_permission=NO_PERMISSIONS`, as you noted.

🤖 *Claude response*
